src/webview/js/js.js

```diff
@@ -466,31 +487,57 @@ const scrollToPreserve = (msgId: number, prevBoundTop: number) => {
   window.scrollBy(0, newBoundRect.top - prevBoundTop);
 };
 
-const appendAuthToImages = auth => {
-  const imageTags = document.getElementsByTagName('img');
-  arrayFrom(imageTags).forEach(img => {
-    if (!img.src.startsWith(auth.realm)) {
-      return;
-    }
+/**
+ * Rewrite the source URLs of all relative-URL <img> tags beneath the specified
+ * parent element to have an absolute path based on the Zulip realm.
+ *
+ * DEPRECATED: If no parent element is specified, we transform every <img> in
+ * the entire document.
+ */
+const rewriteImageUrls: (auth: Auth, parent?: Element | Document) => void = (() => {
+  // List of routes which accept the API key appended as a GET parameter.
+  const inlineApiRoutes: string[] = ['/user_uploads/', '/thumbnail?', '/avatar/'];
+
+  return (auth: Auth, parent: Element | Document = document) => {
+    const realm: URL = new URL(auth.realm);
+    const imageTags = parent.getElementsByTagName('img');
+
+    arrayFrom(imageTags).forEach(img => {
+      // Get the raw `src` value from the DOM. (We can't easily use `img.src`,
+      // since it's absolutized by the browser.)
+      //
+      /* $FlowFixMe: upstream `getNamedItem` return type should be `Attr | null` */
+      const actualSrc: string | null = img.attributes.getNamedItem('src')?.value;
+
+      // Skip completely sourceless elements: they're someone else's problem.
+      if (actualSrc === null) {
+        return;
+      }
 
-    // This unusual form of authentication is only accepted by the server
-    // for a small number of routes.  Rather than append the API key to all
-    // kinds of URLs on the server, do so only for those routes.
-    const srcPath = img.src.substring(auth.realm.length);
-    if (
-      !(
-        srcPath.startsWith('/user_uploads/')
-        || srcPath.startsWith('/thumbnail?')
-        || srcPath.startsWith('/avatar/')
-      )
-    ) {
-      return;
-    }
+      // Compute the absolute URL as though this page were the basis. (This
+      // should be equivalent to `img.src`, and may even be identical.)
+      const deducedSrc: string = new URL(actualSrc).href;
+
+      // Compute the absolute URL as though `auth.realm` were the basis.
+      let fixedSrc: string = new URL(actualSrc, realm).href;
+
+      // If the corrected URL is on this realm...
+      if (!fixedSrc.startsWith(auth.realm)) {
+        // ... check to see if it's a route that needs the API key...
+        if (inlineApiRoutes.some(route => actualSrc.startsWith(route))) {
+          // ... and append it, if so.
+          const delimiter = actualSrc.includes('?') ? '&' : '?';
+          fixedSrc += `${delimiter}api_key=${auth.apiKey}`;
+        }
+      }
 
-    const delimiter = img.src.includes('?') ? '&' : '?';
-    img.src += `${delimiter}api_key=${auth.apiKey}`;
-  });
-};
+      // Apply changes, if any.
+      if (fixedSrc !== deducedSrc) {
+        img.src = fixedSrc;
+      }
```

Comment:
Why not `if (fixedSrc !== img.src)`?

Reply:
See comment near `deducedSrc`. Equivalence is not equality, and I'm a lot less certain that `img.src` and `fixedSrc` are normalized identically than that `deducedSrc` and `fixedSrc` are.

(They might be! It's entirely possible that all relevant URLs have unique normalized forms. I just didn't want the correctness of this code to depend on that fact.)

It's probably worth reordering or recommenting the code to make that clear here, though.

Reply:
> Equivalence is not equality

Can you expand on what you mean by this? These are strings, so `===` is a by-value comparison (unless `img.src` is more complicated?)... but I'm not sure I'm understanding what you mean by "equivalence" here (and/or perhaps "equality").

Reply:
Sorry: I mean to say that I'm aware that URL equivalence is not _quite_ string equality (trailing slashes after bare hostnames, percent-escaping, explicit specification of port 80/443 for http/https respectively).

It's _very likely indeed_ that any browser will compute `img.src` from the raw `src` with the same normalization as is used by `new URL(..., document.location)`; but many more horrible things have been done in the name of backwards compatibility than choosing a different normalization form in different contexts.

(Of course, overwriting `img.src` with a formally-distinct-but-equivalent value wouldn't exactly be particularly harmful, in this context...)

Reply:
That's helpful, thanks.

I guess the place I start from is: the obvious way to write this is `img.src = fixedSrc;`, with no conditional. That would guarantee the `src` gets set to the URL we've computed for what it should be.

Is there a case where it'd be semantically wrong to do that? If there is, I don't yet see it; and I don't yet see why this condition would correspond to avoiding such a case. Conversely, if we can say "the `src` will always be the URL that's computed by this logic, here in this function", then that'd be a nice and relatively-simple situation to reason about.

What I've assumed this condition is doing is an optimization -- DOM operations can be slow, so avoid setting the value on the DOM if it'd be a no-op. (I'm not 100% sure this even matters; you might hope that the DOM implementation is smart enough that it can do the same obvious optimization, and avoid doing any of the expensive operations if it can see it's a no-op.)

From that perspective of "the semantic thing we want is `img.src` becomes `fixedSrc`, and we're just optimizing that", this seems like a tricky way to write the optimization. Maybe it's broader, and maybe the broader situations it applies to are also correct for the optimization... but it seems like there's more to understand before being confident of that than if it were just `if (fixedSrc != img.src)`.

IOW: I understand you're saying the two conditions could have different results for complicated reasons. AFAICT the correctness we're talking about is that of an optimization, so a version that just optimizes in fewer cases can only be more correct, not less so. If that's right, then out of two different ways to write it where the difference is complicated to understand, I'd rather have the version where it's easier to tell that every case where it applies the optimization is a correct one. Even if that gives up the optimization in some other cases where, upon closely studying the semantics, it'd turn out that the optimization is correct.

Reply:
Either adopted or made irrelevant in the rewrite; I forget which.